* 异步			：	异步是关于现在和将来的时间间隙，通过时间队列和环境监听将一件事情分为现在和将来两个部分
* 同步操作	：	操作一个一个执行

浏览器中的异步操作有：定时器，事件，promise


### Promise

用同步一样的方式书写异步代码
```
let p = new Promise(function(resolve, reject){
	// 异步代码
	// resolve ->  成功
	// reject -> 失败
	$.ajax({
		url: url,
		dataType: "json",
		success(data){
			resolve();
		},
		error(err){
			reject();
		}
	})
})

p.then(function(data){}, function(res){})

Promise.all([p1, p2]).then(function(dataArr){
	let [res1, res2] = dataArr;
}, function(res){
})
```

实际上 jquery 的 ajax 方法返回的就是一个 promise 对象
```
Promise.all([
	$.ajax({url: "array.txt", dataType: "json"}),
	$.ajax({url: "json.txt", dataType: "json"})
]).then(function(results){
	let [res1, res2] = results;
}, function(error){
})
```

##### any
Promise.any()  (Firefox > 79, Chrome > 85)  接收一个 Promise 可迭代对象，只要其中的一个 promise 成功，就返回那个已经成功的 promise 。如果可迭代对象中没有一个 promise 成功（即所有的 promises 都失败/拒绝），就返回一个失败的 promise 和AggregateError类型的实例，它是 Error 的一个子类，用于把单一的错误集合在一起。本质上，这个方法和Promise.all()是相反的。

##### race

和 Promise.all 不同，all 是要加入 Promise 数组中的所有对象都执行完了才算成功，有一个失败就算失败了；
Promise.race  是其中一个 Promise 对象执行成功就算成功了，其它的就不执行。

### generator

普通函数执行的时候是从开始一直执行到函数返回
generator 也是一个函数，不过比较特殊的是执行过程中可以暂停，让程序执行其它的东西
声明 generator 函数的方式, 不能简写成箭头函数
```
function *foo(){
	alert('a');

	yield;

	alert('b');
}

let generatorObj = foo();
generatorObj.next(); // 执行到第一个 yield, 没有 yield 就执行到函数结尾
```
##### yield

生成器函数的返回值是一个迭代器对象(iterator object)。要从生成器中提取值，我们可以使用两种方法:

- 使用 next() 方法
- 通过 for...of 遍历

+ 传参
+ 有返回值

返回值示例： next
```
function *foo(){
	alert('a');

	yield 12;  //  返回值

	alert('b');
	return 5;
}

let obj = foo();
let ret1 = obj.next(); // ret1 = {value: 12, done: false}
let ret2 = obj.next(); // ret2 = {value: 5, done: true}
```

##### for of
```
function* generate() {
  yield 33;
  yield 99;

  throw Error("我要结束你!")
}

try {
  for (const value of generate()) {
    console.log(value)
  }
} catch (error) {
  console.log(error.message)
}

/* 输出：
  33
  99
  我要结束你!
*/
```

传参示例：
```
function *foo(a){
	alert(a);

	let c = yield;

	alert(c);
}

let obj = foo(9);
obj.next();   // a = 9
obj.next(90); // c = 90
```

###### 除了next()之外，从生成器返回的迭代器对象还具有throw()方法。使用这种方法，我们可以通过向生成器中注入一个异常来停止程序
```
function* generate() {
  yield 33;
  yield 99;
}

const go = generate();

const firstStep = go.next().value; // 33

go.throw(Error("我要结束你!"));

const secondStep = go.next().value; // 这里会抛出异常
```

使用场景：当执行了一部分时需要等待其它操作结束

npm i yield-runner-blue

### async & await

async await 取代了 generator
场景：三个操作a1-a3, a2需要使用 a1 的返回结果
使用 promise 是不行的，没有顺序控制。generator 可以做到，但是 async 和 await 更加优雅。
```
async function show(){
	let promise1 = $.ajax(url: "data1.json", dataType: "json");
	let promise2 = $.ajax(url: "data2.json", dataType: "json");
	let promise3 = $.ajax(url: "data3.json", dataType: "json");
	let data1 = await promise1;
	if(data1){
		let data2 = await promise2;
	}else{
		let data3 = await promise3;
	}
}
```

> https://www.bilibili.com/video/av50680998/?p=138
